src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/DependencyInjectionSpecificationTests.cs

```diff
@@ -122,36 +122,91 @@ public void TransientServiceCanBeResolvedFromScope()
+                otherFakeServiceFromScope2 = serviceWithProvider.ServiceProvider.GetRequiredService<IFakeService>();
+            }
+
+            // Assert
+            Assert.Same(fakeServiceFromScope1, otherFakeServiceFromScope1);
+            Assert.Same(fakeServiceFromScope2, otherFakeServiceFromScope2);
+            Assert.NotSame(fakeServiceFromScope1, fakeServiceFromScope2);
+        }
+
         [Fact]
         public void SingletonServiceCanBeResolvedFromScope()
         {
             // Arrange
             var collection = new TestServiceCollection();
+            // When resolved from the root service provider / root scope, the scoped IFakeService effectively
+            // also becomes a singleton.
+            collection.AddScoped<IFakeService, FakeService>();
             collection.AddSingleton<ClassWithServiceProvider>();
             var provider = CreateServiceProvider(collection);
 
             // Act
-            IServiceProvider scopedSp1 = null;
-            IServiceProvider scopedSp2 = null;
-            ClassWithServiceProvider instance1 = null;
-            ClassWithServiceProvider instance2 = null;
+            IFakeService fakeServiceFromScope1 = null;
+            IFakeService scopedFakeServiceFromScope1 = null;
+            IFakeService fakeServiceFromScope2 = null;
+            IFakeService scopedFakeServiceFromScope2 = null;
 
             using (var scope1 = provider.CreateScope())
             {
-                scopedSp1 = scope1.ServiceProvider;
-                instance1 = scope1.ServiceProvider.GetRequiredService<ClassWithServiceProvider>();
+                scopedFakeServiceFromScope1 = scope1.ServiceProvider.GetRequiredService<IFakeService>();
+                var serviceWithProvider = scope1.ServiceProvider.GetRequiredService<ClassWithServiceProvider>();
+                // resolved through singleton, the fake service should be scoped to root scope
+                fakeServiceFromScope1 = serviceWithProvider.ServiceProvider.GetRequiredService<IFakeService>();
             }
 
             using (var scope2 = provider.CreateScope())
             {
-                scopedSp2 = scope2.ServiceProvider;
-                instance2 = scope2.ServiceProvider.GetRequiredService<ClassWithServiceProvider>();
+                scopedFakeServiceFromScope2 = scope2.ServiceProvider.GetRequiredService<IFakeService>();
+                var serviceWithProvider = scope2.ServiceProvider.GetRequiredService<ClassWithServiceProvider>();
+                // resolved through singleton, the fake service should be scoped to root scope
+                fakeServiceFromScope2 = serviceWithProvider.ServiceProvider.GetRequiredService<IFakeService>();
             }
 
+            IFakeService fakeServiceFromRootScope = provider.GetRequiredService<IFakeService>();
+
             // Assert
-            Assert.Same(instance1.ServiceProvider, instance2.ServiceProvider);
```

Comment:
I think this is breaking the original intention of this test. See https://github.com/dotnet/extensions/pull/2236 and https://github.com/dotnet/extensions/issues/1301 for more information on why this test was written this way.

Reply:
If I actually _enable_ that singleton test for Unity (where it is currently still disabled), then it does still _fail_ that test, but in an "interesting" way:
```
Failed Microsoft.Extensions.DependencyInjection.Specification.UnityDependencyInjectionSpecificationTests.SingletonServiceCanBeResolvedFromScope [23 ms]
  Error Message:
   System.InvalidOperationException : No service for type 'Microsoft.Extensions.DependencyInjection.Specification.Fakes.IFakeService' has been registered.
  Stack Trace:
     at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType) in /workspaces/runtime/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceProviderServiceExtensions.cs:line 58
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider) in /workspaces/runtime/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceProviderServiceExtensions.cs:line 78
   at Microsoft.Extensions.DependencyInjection.Specification.DependencyInjectionSpecificationTests.SingletonServiceCanBeResolvedFromScope() in /workspaces/runtime/src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/DependencyInjectionSpecificationTests.cs:line 196
```

Meaning that when it tries to resolve the fake service in the **second** scope, it can't find a registration for `IFakeService`. This is admittedly _not_ what I was expecting.

Reply:
Just to be clear. I'll of course revert the changes to this test if you prefer, but I do think that the reference equality testing on the service provider is unnecessary and misleading as demonstrated by my two simple demo tests above.

Reply:
> as demonstrated by my two simple demo tests above.

This test is for Singleton services, and both those demo tests use Scoped services.

For Singleton services, I believe the intention is that the ServiceProvider they get _is_ the root provider, and thus they should be the same.

@davidfowl - you originally wrote the test. Do you have any thoughts here?

Reply:
Yes, this test is for Singleton services. And the `ClassWithServiceProvider` is still registered as a singleton and thus should be injected with the root provider. The scoped `IFakeService` is just my "canary in the coal mine" for figuring out if the singleton was actually injected with the root scope. It still tests the same scenario, just in a different way.

* `scopedFakeServiceFromScope1` is resolved from the scope's service provider => that should be a fresh instance bound to `scope1`
* `fakeServiceFromScope1` is resolved from the service provider that was injected into the singleton which **should** be the root scope => that should give me a different instance to `scopedFakeServiceFromScope1`
* then we play the same game with `scope2` and compare results based on where we expect the same or different instances

Reply:
Let me take a deeper look at this today. On second glance I can see why my test would make other containers fail. The service provider instance isn't as important as the lifetime of the objects that come out of it.